src/java/org/apache/cassandra/index/sai/SSTableContextManager.java

```diff
@@ -0,0 +1,148 @@
+
+    /**
+     * Initialize {@link SSTableContext}s if they are not already initialized.
+     *
+     * @param removed SSTables being removed
+     * @param added SSTables being added
+     * @param validate if true, header and footer will be validated.
+     *
+     * @return a set of contexts for SSTables with valid per-SSTable components, and a set of
+     * SSTables with invalid or missing components
+     */
+    public Pair<Set<SSTableContext>, Set<SSTableReader>> update(Collection<SSTableReader> removed, Iterable<SSTableReader> added, boolean validate)
+    {
+        release(removed);
+
+        Set<SSTableContext> contexts = new HashSet<>();
+        Set<SSTableReader> invalid = new HashSet<>();
+
+        for (SSTableReader sstable : added)
+        {
+            if (sstable.isMarkedCompacted())
+            {
+                continue;
+            }
+
+            IndexDescriptor indexDescriptor = IndexDescriptor.create(sstable);
+
+            if (!indexDescriptor.isPerSSTableIndexBuildComplete())
+            {
+                // Don't even try to validate or add the context if the completion marker is missing.
+                continue;
+            }
+
+            try
+            {
+                // Only validate on restart or newly refreshed SSTable. Newly built files are unlikely to be corrupted.
+                if (validate && !sstableContexts.containsKey(sstable) && !indexDescriptor.validatePerSSTableComponents())
+                {
+                    logger.warn(indexDescriptor.logMessage("Invalid per-SSTable component for SSTable {}"), sstable.descriptor);
+                    invalid.add(sstable);
+                    removeInvalidSSTableContext(sstable);
+                    continue;
+                }
+                // ConcurrentHashMap#computeIfAbsent guarantees atomicity, so {@link SSTableContext#create(SSTableReader)}}
+                // is called at most once per key.
+                contexts.add(sstableContexts.computeIfAbsent(sstable, SSTableContext::create));
+            }
+            catch (Throwable t)
+            {
+                logger.warn(indexDescriptor.logMessage("Failed to update per-SSTable components for SSTable {}"), sstable.descriptor, t);
+                invalid.add(sstable);
+                removeInvalidSSTableContext(sstable);
+            }
+        }
+
+        return Pair.create(contexts, invalid);
+    }
+
+    public void release(Collection<SSTableReader> toRelease)
+    {
+        toRelease.stream().map(sstableContexts::remove).filter(Objects::nonNull).forEach(SSTableContext::close);
+    }
+
+    /**
+     * @return total number of per-sstable open files for live sstables
+     */
+    int openFiles()
+    {
+        return sstableContexts.values().stream().mapToInt(SSTableContext::openFilesPerSSTable).sum();
+    }
+
+    /**
+     * @return total disk usage of all per-sstable index files
```

Comment:
Nit: this could mention the unit
```suggestion
     * @return total disk usage (in bytes) of all per-sstable index files
```